### PR TITLE
delta xds: never remove ECDS explicitly

### PR DESCRIPTION
**Pull request by howardjohn**
Fixes https://github.com/istio/istio/issues/49847

### CODE REVIEW

File: pilot/pkg/xds/delta.go [hunk 1]
```diff
@@ -502,6 +502,9 @@ func (s *DiscoveryServer) pushDeltaXds(con *Connection,
 		removed := subscribed.DeleteAll(currentResources...)
 		resp.RemovedResources = sets.SortedList(removed)
 	}
+	if neverRemoveDelta(w.TypeUrl) {
```
Comment (keithmattix) [inline pilot/pkg/xds/delta.go:505]: nit: add Debug line for now since this isn't intuitive behavior

File: pilot/pkg/xds/delta.go [hunk 2]
```diff
@@ -561,6 +564,13 @@ func requiresResourceNamesModification(url string) bool {
 	return url == v3.AddressType || url == v3.WorkloadType
 }
 
+// neverRemoveDelta checks if a type should never remove resources
+func neverRemoveDelta(url string) bool {
+	// https://github.com/envoyproxy/envoy/issues/32823
+	// We want to garbage collect extensions when they are no longer referenced, rather than delete immediately
+	return url == v3.ExtensionConfigurationType
```
Comment (keithmattix) [inline pilot/pkg/xds/delta.go:571]: That issue mentions other top-level types; should we never remove those either?

Comment (howardjohn) [reply pilot/pkg/xds/delta.go:571]: ECDS is not a top level resource, only CDS and LDS are. The rest of the dependant resources (that we use in Istio at least) do not handle deletes like ECDS. Instead delete basically means "You requested this and we don't have it".

Maybe we still need a way to express that for ECDS though...

Comment (keithmattix) [reply pilot/pkg/xds/delta.go:571]: Not sure I understand this formulation; isn't the crux of the issue that ECDS deletes are applied while listener is draining?

Comment (howardjohn) [reply pilot/pkg/xds/delta.go:571]: So consider a route, which is also tied to a listener. If you send "Remove route X" that is still referenced by a (presumably, draining) listener, it is entirely ignored (_any_ "Remove" is ignored, its not just about draining).
SDS works the same


ECDS is different, and instantly deletes it causing 500 errors (NFCF) for requests during the drain

Comment (keithmattix) [reply pilot/pkg/xds/delta.go:571]: Ack - in that case I think this is fine; we'll just assume that the removal is surfaced via the listener removing it

**Review by keithmattix — APPROVED** [review]

**Comment by howardjohn** [comment]
I put a hold since I am not confident this is a good idea.

cc @zirain @hanxiaop @kyessenov @ramaraochavali @hzxuzhonghu

**Comment by kyessenov** [comment]
It's not a good idea :) However, this is only impacting Wasm AFAIK which is already in a pretty broken state.

**Comment by zirain** [comment]
> It's not a good idea :) However, this is only impacting Wasm AFAIK which is already in a pretty broken state.

yeah, it's the only thing we can do.

**Review by zirain — APPROVED** [review]

**Pull request merged**